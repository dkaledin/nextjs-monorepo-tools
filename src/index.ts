import path from "path";
import fs from "fs";
import { Configuration, Resolve } from "webpack";

export interface PatchWebpackConfigParams {
  commonDirs: string[];
}
export type TsPaths = { [key: string]: string[] };

export function patchWebpackConfig({ commonDirs }: PatchWebpackConfigParams) {
  try {
    const tsConfigPath = findFilepath("tsconfig.json");
    const tsConfigJson = fs.readFileSync(tsConfigPath, { encoding: "utf-8" });
    const tsConfig = jsonParse(tsConfigJson);
    const tsPaths = tsConfig.compilerOptions.paths;

    if (!tsPaths) {
      console.error("You should specify paths aliases in tsconfig.json");
      process.exit(1);
    }

    return (config: Configuration): Configuration => {
      const alias: Resolve["alias"] = Object.keys(tsPaths).reduce(
        (result, key) => ({
          ...result,
          [key]: tsPaths[key][0],
        }),
        {}
      );
      const patchedConfig = {
        ...config,
        resolve: {
          ...config.resolve,
          alias: {
            ...config.resolve?.alias,
            ...alias,
          },
        },
      };

      const resolverPathsOfCommonDirectories = commonDirs.map<string>((dir) =>
        path.resolve(process.cwd(), dir)
      );

      ((patchedConfig.module?.rules[0]?.include as string[]) || []).push(
        ...resolverPathsOfCommonDirectories
      );

      return patchedConfig;
    };
  } catch (error) {
    console.error(error);
    process.exit(1);
  }
}

let index = 0;

function findFilepath(filename: string, currentPath: string[] = []): string {
  ++index;
  const resolvedPath = path.resolve(process.cwd(), ...currentPath, filename);

  console.log("resolvedPath", resolvedPath);

  if (index > 10) {
    return "fail_path";
  }

  if (fs.existsSync(resolvedPath)) {
    return resolvedPath;
  } else {
    return findFilepath(filename, [...currentPath, "../"]);
  }
}

function jsonParse(json: string): any {
  return JSON.parse(json);
}
