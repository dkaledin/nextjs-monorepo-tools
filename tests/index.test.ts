import { Configuration } from "webpack";
import { patchWebpackConfig } from "../src";

it("should return expected Webpack configureation", () => {
  const nextjsDefaultConfig = {
    resolve: {
      modules: ["node_modules"],
      alias: {
        "next/head": "next/dist/next-server/lib/head.js",
        "next/router": "next/dist/client/router.js",
        "next/config": "next/dist/next-server/lib/runtime-config.js",
      },
    },
    module: {
      rules: [
        {
          test: /\.(tsx|ts|js|mjs|jsx)$/,
          include: [
            "/example-user-directory/next-project/apps/nest-app",
            /next[\\/]dist[\\/]next-server[\\/]lib/,
            /next[\\/]dist[\\/]client/,
            /next[\\/]dist[\\/]pages/,
            /[\\/](strip-ansi|ansi-regex)[\\/]/,
          ],
          exclude: [],
          use: { loader: "next-babel-loader", options: [Object] },
        },
        {
          oneOf: [[Object], [Object], [Object]],
        },
      ],
      strictExportPresence: true,
    },
  };
  const actual = patchWebpackConfig({
    commonDirs: ["libs"],
  })(nextjsDefaultConfig as Configuration);

  expect(actual).toMatchSnapshot();
});
