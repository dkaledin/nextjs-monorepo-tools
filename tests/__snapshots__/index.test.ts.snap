// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return expected Webpack configureation 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "exclude": Array [],
        "include": Array [
          "/example-user-directory/next-project/apps/nest-app",
          /next\\[\\\\\\\\/\\]dist\\[\\\\\\\\/\\]next-server\\[\\\\\\\\/\\]lib/,
          /next\\[\\\\\\\\/\\]dist\\[\\\\\\\\/\\]client/,
          /next\\[\\\\\\\\/\\]dist\\[\\\\\\\\/\\]pages/,
          /\\[\\\\\\\\/\\]\\(strip-ansi\\|ansi-regex\\)\\[\\\\\\\\/\\]/,
          "/Users/g0bl1n/projects/nextjs-monorepo-tools/libs",
        ],
        "test": /\\\\\\.\\(tsx\\|ts\\|js\\|mjs\\|jsx\\)\\$/,
        "use": Object {
          "loader": "next-babel-loader",
          "options": Array [
            [Function],
          ],
        },
      },
      Object {
        "oneOf": Array [
          Array [
            [Function],
          ],
          Array [
            [Function],
          ],
          Array [
            [Function],
          ],
        ],
      },
    ],
    "strictExportPresence": true,
  },
  "resolve": Object {
    "alias": Object {
      "@foo/bar": "foo/bar",
      "next/config": "next/dist/next-server/lib/runtime-config.js",
      "next/head": "next/dist/next-server/lib/head.js",
      "next/router": "next/dist/client/router.js",
    },
    "modules": Array [
      "node_modules",
    ],
  },
}
`;
